crates/rustc_codegen_spirv/src/abi.rs

```diff
@@ -705,3 +712,49 @@ fn name_of_struct(ty: TyAndLayout<'_>) -> String {
     }
     name
 }
+
+fn trans_image<'tcx>(
+    cx: &CodegenCx<'tcx>,
+    ty: TyAndLayout<'tcx>,
+    attr: SpirvAttribute,
+) -> Option<Word> {
+    match attr {
+        SpirvAttribute::Image {
+            dim,
+            depth,
+            arrayed,
+            multisampled,
+            sampled,
+            image_format,
+            access_qualifier,
+        } => {
+            // see SpirvType::sizeof
+            if ty.size != Size::from_bytes(4) {
+                cx.tcx.sess.err("#[spirv(image)] type must have size 4");
+                return None;
+            }
```

Comment:
Just leaving this as a note, my preference is encoding it as `&Image` where
```rust
extern {
    #[spirv(image)]
    type Image;
}
```

But I have no idea if this is workable, so we should leave it for future (if ever) experimentation.

<hr/>

With a hardcoded size, one fun way to stop `rustc` from letting you nest the layout in anything else (with any more data), is to make it a newtype of `[u8; (1 << 47) - 1]`. That's because you can't have a size of 2<sup>47</sup> (or larger), on a 64-bit target (and there's a similar limitation on 32-bit targets but it's closer to 2<sup>32</sup>).

Reply:
The other way to think about this is we could represent them as small as `1`-byte structs, *but no less*, so that the offset bijectively maps to the field, and generally to avoid any ZST special-casing.

The only reason to use pointer-size is the analogy with WASM, which has e.g. "function pointers" (and more generally `externref`), that are abstract values which can be referred to by an `usize` index into a global "table", if need be.
(Is that a plausible concern/direction with SPIR-V? Could we actually have a global array of e.g. `Image`s?)

We can also avoid any `Abi` special-casing by using `[u8; 1]`, IIRC, since that *should* create `Abi::Aggregate`.
(Either way we should assert the `.abi` is what we expect, in case it *does* change in the future)